rust/primary/src/header_waiter.rs

```diff
@@ -0,0 +1,292 @@
+                            let round = header.round;
+                            let author = header.author;
+
+                            // Ensure we sync only once per header.
+                            if self.pending.contains_key(&header_id) {
+                                continue;
+                            }
+
+                            // Add the header to the waiter pool. The waiter will return it to when all
+                            // its parents are in the store.
+                            let wait_for = missing
+                                .iter()
+                                .map(|(digest, worker_id)| {
+                                    let key = [digest.as_ref(), &worker_id.to_le_bytes()].concat();
+                                    (key.to_vec(), self.store.clone())
+                                })
+                                .collect();
+                            let (tx_cancel, rx_cancel) = channel(1);
+                            self.pending.insert(header_id, (round, tx_cancel));
+                            let fut = Self::waiter(wait_for, header, rx_cancel);
+                            waiting.push(fut);
+
+                            // Ensure we didn't already send a sync request for these parents.
+                            let mut requires_sync = HashMap::new();
+                            for (digest, worker_id) in missing.into_iter() {
+                                self.batch_requests.entry(digest.clone()).or_insert_with(|| {
+                                    requires_sync.entry(worker_id).or_insert_with(Vec::new).push(digest);
+                                    round
+                                });
+                            }
+                            for (worker_id, digests) in requires_sync {
+                                let address = self.committee
+                                    .worker(&author, &worker_id)
+                                    .expect("Author of valid header is not in the committee")
+                                    .primary_to_worker;
+                                let message = PrimaryWorkerMessage::Synchronize(digests, author);
+                                let bytes = bincode::serialize(&message)
+                                    .expect("Failed to serialize batch sync request");
+                                self.network.send(address, Bytes::from(bytes)).await;
+                            }
+                        }
+
+                        WaiterMessage::SyncParents(missing, header) => {
+                            debug!("Synching the parents of {}", header);
+                            let header_id = header.id.clone();
+                            let round = header.round;
+                            let author = header.author;
+
+                            // Ensure we sync only once per header.
+                            if self.pending.contains_key(&header_id) {
```

Comment:
IOptional feature: f we get the same parent from more than one peers do we want to be smart and ask these peers first before random selection?

Reply:
Updated issue #10